Core/Database.swift

```diff
@@ -0,0 +1,112 @@
+//
+//  Database.swift
+//  Core
+//
+//  Copyright © 2019 DuckDuckGo. All rights reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+import CoreData
+
+public class Database {
+    
+    fileprivate struct Constants {
+        static let databaseGroupID = "group.com.duckduckgo.database"
+        
+        static let databaseName = "Database"
+    }
+    
+    public static let shared = Database()
+
+    private let semaphore = DispatchSemaphore(value: 0)
+    private let container: NSPersistentContainer
+    
+    public var model: NSManagedObjectModel {
+        return container.managedObjectModel
+    }
+    
+    convenience init() {
+        let mainBundle = Bundle.main
+        let coreBundle = Bundle(identifier: "com.duckduckgo.mobile.ios.Core")!
+        
+        guard let managedObjectModel = NSManagedObjectModel.mergedModel(from: [mainBundle, coreBundle]) else { fatalError("No DB scheme found") }
+        
+        self.init(name: Constants.databaseName, model: managedObjectModel)
+    }
+    
+    init(name: String, model: NSManagedObjectModel) {
+        container = DDGPersistentContainer(name: name, managedObjectModel: model)
+    }
+    
+    public func loadStore(andMigrate handler: @escaping (NSManagedObjectContext) -> Void = { _ in }) {
+        container.loadPersistentStores { _, error in
+            if let error = error {
+                Pixel.fire(pixel: .dbInitializationError, error: error)
+                // Give Pixel a chance to be sent, but not too long
+                Thread.sleep(forTimeInterval: 1)
```

Comment:
`Pixel.fire` has a completion handler you could use instead - would that be better?

Reply:
I considered that, but:
- I don't want to block this code for too long - just crash if needed.
- I'd need to block the thread here and wait for completion to execute.
- Pixel.fire's completion block is queued on main thread - so if there is already something on main thread waiting for context (and thus - blocking main thread till DB is ready) it could result in a deadlock. I'd have to change Pixel.fire API to return on private queue to make sure that works.

Reply:
Makes sense